src/common/summarize-profile.js

```diff
@@ -108,23 +94,51 @@ export function summarizeProfile(profile: Profile) {
  * are cached between calls.
  * @returns {function} Function categorizer.
  */
-function functionNameCategorizer() {
-  const cache = new Map();
-  return function functionNameToCategory(name) {
-    const existingCategory = cache.get(name);
-    if (existingCategory !== undefined) {
+function functionNameCategorizer(numFunctions): (number, number[], string[]) => string {
+  const cache = new Array(numFunctions);
+  cache.fill('', 0, numFunctions);
+
+  // Compile set of categories into a single regular expression that produces
+  // the pattern (whether it's a prefix, stem, substring, or exact match).
+
+  const patternToCategory = new Map();
+
+  const matchPatterns = {
+    prefix: [],
+    stem: [],
+    substring: [],
+    exact: [],
+  };
+
+  for (const [matchType, pattern, category] of categories) {
+    patternToCategory.set(pattern, category);
```

Comment:
So this assumes that there are no collisions of two matchers with the same pattern (but different matchTypes and different categories). That's probably a reasonable assumption, but I think it needs to be written down somewhere.